src/lib/stores.ts

```diff
@@ -0,0 +1,130 @@
+import { writable, derived } from 'svelte/store';
+import { withoutTransition } from './without-transition';
+
+// saves having to branch for server vs client
+const noopStorage = {
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	getItem: (_key: string) => null,
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	setItem: (_key: string, _value: string) => {}
+};
+
+/**
+ * The key used to store the mode in localStorage.
+ */
+export const localStorageKey = 'mode';
+/**
+ * Writable store that represents the user's preferred mode (`"dark"`, `"light"` or `"system"`)
+ */
+export const userPrefersMode = createUserPrefersMode();
+/**
+ * Readable store that represents the system's preferred mode (`"dark"`, `"light"` or `undefined`)
+ */
+export const systemPrefersMode = createSystemMode();
+/**
+ * Derived store that represents the current mode (`"dark"`, `"light"` or `undefined`)
+ */
+export const derivedMode = createDerivedMode();
+
+// derived from: https://github.com/CaptainCodeman/svelte-web-storage
+function createUserPrefersMode() {
+	const browser = typeof window !== 'undefined' && typeof document !== 'undefined';
+	const defaultValue = 'system';
+
+	const storage = browser ? localStorage : noopStorage;
+	let value = (storage.getItem(localStorageKey) as 'dark' | 'light' | 'system') || defaultValue;
+
+	const { subscribe, set: _set } = writable(value, () => {
+		if (browser) {
+			const handler = (e: StorageEvent) => {
+				if (e.key === localStorageKey) {
+					_set((value = (e.newValue as 'dark' | 'light' | 'system') || defaultValue));
+				}
+			};
+			addEventListener('storage', handler);
+			return () => removeEventListener('storage', handler);
+		}
+	});
+
+	function set(v: 'dark' | 'light' | 'system') {
+		_set((value = v));
+		storage.setItem(localStorageKey, value);
+	}
+
+	return {
+		subscribe,
+		set
+	};
+}
+
+function createSystemMode() {
+	const browser = typeof window !== 'undefined' && typeof document !== 'undefined';
```

Comment:
Same as above, I think we could just move this outside the scope of the functions tbh:

```ts
const isBrowser = typeof document !== undefined
```
and then reuse it throughout.

Reply:
sounds good!